### PR TITLE
Add Pip installation report experimental detector

### PR DESCRIPTION
**Pull request by cobya**
As a follow up to #108 and #980, add experimental Python detection via `pip install -r requirements.txt --dry-run --ignore-installed --quiet --report {reportName}` or equivalent.

The --report option of the pip install command produces a detailed JSON report of what it did install (or what it would have installed. See https://pip.pypa.io/en/stable/reference/installation-report/#specification for more details.

Serialization specifications:
- https://packaging.python.org/en/latest/specifications/core-metadata/
- https://peps.python.org/pep-0508/
- https://peps.python.org/pep-0301/

### CODE REVIEW

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/IPipCommandService.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+using System.IO;
+using System.Threading.Tasks;
+
+public interface IPipCommandService
+{
+    Task<bool> PipExistsAsync(string pipPath = null);
+
+    Task<(PipInstallationReport Report, FileInfo ReportFile)> GenerateInstallationReportAsync(string path, string pipExePath = null);
+}
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/Contracts/IPipCommandService.cs:11]: nit: docs

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+using Newtonsoft.Json;
+
+/// <summary>
+/// Metadata for a pip component being installed. See https://packaging.python.org/en/latest/specifications/core-metadata/.
+/// Some fields are not collected here because they are not needed for dependency graph construction.
+/// </summary>
+public class PipInstallationMetadata
+{
+    /// <summary>
+    /// Version of the file format; legal values are “1.0”, “1.1”, “1.2”, “2.1”, “2.2”, and “2.3”
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs:12]: nit: are these smart quotes?

```suggestion
    /// Version of the file format; legal values are "1.0", "1.1", "1.2", "2.1", "2.2", and "2.3"
```

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
+namespace Microsoft.ComponentDetection.Detectors.Pip;
+
+using Newtonsoft.Json;
+
+/// <summary>
+/// Metadata for a pip component being installed. See https://packaging.python.org/en/latest/specifications/core-metadata/.
+/// Some fields are not collected here because they are not needed for dependency graph construction.
+/// </summary>
+public class PipInstallationMetadata
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs:9]:
```suggestion
public sealed record PipInstallationMetadata
```

File: src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            return pipCommand;
+        }
+
+        return null;
+    }
+
+    private async Task<bool> CanCommandBeLocatedAsync(string pipPath)
+    {
+        return await this.commandLineInvocationService.CanCommandBeLocatedAsync(pipPath, new List<string> { "pip3" }, "--version");
+    }
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs:57]: Need to do a version check here? Installation report was only added in pip 22.2[^1]

[^1]: https://pip.pypa.io/en/stable/news/#v22-2

File: src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        else if (path.EndsWith(".txt"))
+        {
+            pipReportCommand = "install -r requirements.txt";
+        }
+        else
+        {
+            // Failure case, but this shouldn't be hit since detection is only running
+            // on setup.py and requirements.txt files.
+            return (new PipInstallationReport(), null);
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs:87]: Log this?

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+/// <summary>
+/// See https://pip.pypa.io/en/stable/reference/installation-report/#specification.
+/// </summary>
+public class PipInstallationReport
+{
+    /// <summary>
+    /// Version of the installation report specification. Currently 1, but will be incremented if the format changes.
+    /// </summary>
+    [JsonProperty("version")]
+    public int Version { get; set; }
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs:15]: [The documentation][1] says this is a string
>version: the string 1.

[1]: https://pip.pypa.io/en/stable/reference/installation-report/#specification

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    /// Distribution packages (to be) installed.
+    /// </summary>
+    [JsonProperty("install")]
+    public PipInstallationReportItem[] InstallItems { get; set; }
+
+    /// <summary>
+    /// Environment metadata for the report. See https://peps.python.org/pep-0508/#environment-markers.
+    /// </summary>
+    [JsonProperty("environment")]
+    public JObject Environment { get; set; }
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs:33]: nit: This is a Newtonsoft.Json implementation specific detail. Could this be `IDictionary<,>` instead?

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs (hunk 3 above)
Comment (JamieMagee) [reply src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs:9]: Same for all the other new contracts.

Sealed brings significant performance improvements[^1]

[^1]: https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#peanut-butter

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs (hunk 7 above)
Comment (JamieMagee) [reply src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs:33]: Same for all the other contracts that use `JObject`.

File: src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+        var pipExecutable = await this.ResolvePipAsync(pipExePath);
+        var formattedPath = this.pathUtilityService.NormalizePath(path);
+        var workingDir = new DirectoryInfo(this.pathUtilityService.GetParentDirectory(formattedPath));
+
+        CommandLineExecutionResult command;
+        var reportName = Path.GetRandomFileName();
+        var reportFile = new FileInfo(Path.Combine(workingDir.FullName, reportName));
+
+        string pipReportCommand;
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs:74]: nit: `StringBuilder`

File: src/Microsoft.ComponentDetection.Detectors/pip/PipReportComponentDetector.cs [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+                    package.Metadata.Version,
+                    author: PipReportUtilities.GetSupplierFromInstalledItem(package),
+                    license: PipReportUtilities.GetLicenseFromInstalledItem(package)),
+                package.Requested);
+
+            nodeReferences.Add(normalizedPkgName, node);
+
+            // requires_dist will contain information about the dependencies of the package.
+            // However, we don't have PipReportGraphNodes for all dependencies, so we will use
+            // an intermediate layer to store the relathionship and update the graph later.
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/PipReportComponentDetector.cs:154]:
```suggestion
            // an intermediate layer to store the relationships and update the graph later.
```

File: src/Microsoft.ComponentDetection.Detectors/pip/PipReportUtilities.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+using System.Text.RegularExpressions;
+
+internal class PipReportUtilities
+{
+    private const int MaxLicenseFieldLength = 100;
+    private const string ClassifierFieldSeparator = " :: ";
+    private const string ClassifierFieldLicensePrefix = "License";
+
+    /// <summary>
+    /// Normalize the package name format to the stanbard Python Packaging format.
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/pip/PipReportUtilities.cs:13]:
```suggestion
    /// Normalize the package name format to the standard Python Packaging format.
```

**Review by JamieMagee — APPROVED** [review]
Mostly minor comments. Overall, looks good. No blockers from me.

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs (hunk 2 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs]: Yeah it looks like they were 😆 Copied them directly from the documentation.

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs (hunk 3 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationMetadata.cs]: All new contracts converted to sealed records 👍

File: src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs (hunk 5 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs:116]: I'm not concerned on this one - the detector file handling logic should avoid this case.

File: src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs (hunk 6 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pip/Contracts/PipInstallationReport.cs]: Changed it to a string and now compare using `Version` instead

**Pull request merged**